### PR TITLE
return_inferencedata option for pm.sample

### PR DESCRIPTION
**Pull request by michaelosthege**
Most PyMC3 modeling workflows (should) at some point use ArviZ to save and load traces as `InferenceData` objects.

With a `sample(…, return_inferencedata=True/False)` option, the conversion may happen as early as possible, encouraging the user to use ArviZ from the start.

+ [x] wait for next ArviZ release to happen
+ [x] update PR such that
  + [x] pin to latest ArviZ version
  + [x] take out the `NotImplementedError`

+ [x] there are no breaking changes to the user-facing API - but a `FutureWarning` is added, allowing us to default to `return-inferencedata=True` in `v4.0.0` if it turns out to be a good idea
+ [x] in #3883, Oriol made some optimizations to avoid duplicate work of conversion. This PR does the same by doing the conversion already in `sample()` and passing the `idata` for convergence checks.
+ [x] are the changes—especially new features—covered by tests and docstrings? -- test was added, other tests remain untouched, so they assert that the user-facing API did not break
+ [x] consider adding/updating relevant example notebooks - I'll do that on a future PR
+ [x] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

### CODE REVIEW

**Comment by michaelosthege** [comment]
This is the first stage - converting to `InferenceData` earlier.

I'm going to wait for the tests before adding the `return_inferencedata` option.

**Comment by michaelosthege** [comment]
Tests are green.
Now pushing the addition of the `return_inferencedata` kwarg. The tests should remain green...

File: pymc3/sampling.py
Comment (rpgoldman) [inline pymc3/sampling.py]: There's a typo here : `tat [sic] includes ...`

File: pymc3/sampling.py
Comment (rpgoldman) [inline pymc3/sampling.py]: minor points:

1. the `return_inferencedata` should be `bool, optional`
2. this format for docstrings doesn't comply with the numpy format, which has spaces on both sides of the ":"

https://numpydoc.readthedocs.io/en/latest/format.html

**Review by rpgoldman — APPROVED** [review]
Looks good to me, pending tests.

Probably needs some additional tests that exercise returning `InferenceData`.

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: Fixed in my edit commit.

File: pymc3/sampling.py
Comment (rpgoldman) [reply pymc3/sampling.py]: Fixed in my edit commit.

Comment (michaelosthege) [reply pymc3/sampling.py]: thanks 👍

Comment (AlexAndorra) [reply pymc3/sampling.py]: I think I'm the one who deleted the first space before the ":" in a previous PR -- this was because when you double-click on an arg on the website, this annoyingly (and weirdly) selects the arg, the colon and the first word after the colon. 

When the first space is not there (like: `arg: details`), double-clicking only selects the arg, as it should. See [this page](https://arviz-devs.github.io/arviz/generated/arviz.plot_compare.html#arviz.plot_compare) of ArviZ doc for instance, for an example of both (double-click on `figsize` and then on `textsize`). 

So I'm in favor of keeping the previous version, if this doesn't have any negative side effects.

PS: I think that's because of this change that there is a merge conflict: Jon added some clarifications as to how step kwargs can be passed to step methods, and I just merged [his PR](https://github.com/pymc-devs/pymc3/pull/3908).

Comment (rpgoldman) [reply pymc3/sampling.py]: I'm afraid it *does* have a negative effect: Sphinx will not format correctly docstrings that don't have " : ". 

I had put in some of these ": " docstrings myself, but I'm afraid they are wrong.  See the numpy doc I linked above.

Comment (michaelosthege) [reply pymc3/sampling.py]: I have no opinion on this. What does our doc-master, @ColCarroll say?

Comment (rpgoldman) [reply pymc3/sampling.py]: I'll wait for @ColCarroll to weigh in, but I will point out that we agreed to use numpy style, and the numpy style description is quite clear about what it expects.

Here's an example of what Sphinx does wrong if  you don't put the spaces around the colons -- the variable names and types run together:

image

To be honest, though, I can't figure out for sure when Sphinx formats properly and when it does the wrong thing.

**Comment by michaelosthege** [comment]
I've added a test for the new kwarg.

As `.sample(` appears another 78 times in the tests, I'd propose to update those tests with the future PR that changes the default behavior.

So this PR is ready for review - I'll update the release notes when I address your upcoming feedback.

**Comment by michaelosthege** [comment]
You'll notice a `NotImplementedError` when `return_inferencedata==True and discard_tuned_samples==False`.

I'm in contact with Oriol to get `arviz.from_pymc3` support for warmup draws. If this can be arranged in an ArviZ release in the near future, we can use that and pin that ArviZ version for the `3.9` release?

**Comment by michaelosthege** [comment]
I've added a test for the new kwarg.

As `.sample(` appears in 78 other places in the test suite, I'd propose to update those tests on the future PR that changes the default behavior.

So I'd say it's ready for review & I'll add

**Review by AlexAndorra — APPROVED** [review]
Thanks Michael, this looks good to me 👌 
I'll merge tomorrow morning, in case @rpgoldman or @ColCarroll request some last-minute changes.

File: pymc3/sampling.py
Comment (AlexAndorra) [inline pymc3/sampling.py]: Shouldn't we say "In v3.10", as v4.0 will probably be reserved for the first release of PyMC4, as pointed out by @ColCarroll ?

File: pymc3/sampling.py
Comment (AlexAndorra) [inline pymc3/sampling.py]: I think this warning should be deleted, as the ArviZ PR will be merged in the next few days -- and will definitely be merged when this goes out with 3.9

**Review by AlexAndorra — CHANGES_REQUESTED** [review]

File: pymc3/sampling.py
Comment (AlexAndorra) [inline pymc3/sampling.py]: Isn't all that done under the hood by `arviz.from_pymc3` now (cf Oriol's PR)?

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: I will take this out as soon as the ArviZ release number is known - then we can pin on that, or check the version here to maintain compatibility with ArviZ 0.7.0, if anything speaks against pinning `arviz>=0.?.?`

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: Yes it is. Same reason as above.. I'm looking forward to take it out.

File: pymc3/sampling.py
Comment (OriolAbril) [inline pymc3/sampling.py]: There should be an option to set `log_likelihood=True` in case the user wants to calculate loo/waic or whatever with the returned idata object. The hardcoded false made sense when the inference data generated was thrown away. Similar thing would go for tuning/warmup storage, but this should be solved once ArviZ PR warmup is merged

Comment (OriolAbril) [reply pymc3/sampling.py]: Another option would be defaulting to None and set a `data.save_log_likelihood` as ArviZ rcParam for users who want to avoid memory overhead due to log_likelihood data.

File: pymc3/sampling.py
Comment (AlexAndorra) [reply pymc3/sampling.py]: I'd be in favor of pinning latest ArviZ, as this would simplify the code and workflow, and 0.x versions aren't supposed to be "production stable". But I'm of course open to debate.
I'll let know as soon as the ArviZ number is known

File: pymc3/sampling.py
Comment (AlexAndorra) [reply pymc3/sampling.py]: Indeed, this will be solved if we pin latest ArviZ 👌

File: pymc3/sampling.py
Comment (AlexAndorra) [reply pymc3/sampling.py]: +1 for defaulting to None 👍 
People will indeed need the possibility of having the log likelihood in the idata (this is even the default on ArviZ, to compute LOO/WAIC).
I think the easiest way forward would be to hold this PR until ArviZ warmup PR is merged. That way, the different code blocks above can be properly refactored. WDYT?

**Comment by AlexAndorra** [comment]
@michaelosthege, as discussed, @OriolAbril's warmup PR was merged on ArviZ repo, and the next ArviZ release will be 0.8.0, so I think we can reactivate work on this PR and address the comments we made above 👌

**Comment by michaelosthege** [comment]
Yes, we can wait for the ArviZ release.
I'll address the feedback when the release is out.

**Comment by fonnesbeck** [comment]
We should probably either convert one of the example notebooks to using this throughout, or create a new one. The `api_quickstart` is probably the place to put it.

**Comment by AlexAndorra** [comment]
@michaelosthege, ArviZ 0.8.0 hase been released 🍾 
We can reactivate work on this PR and address the comments we made above 👌

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: I've changed it such that it defaults to the ArviZ default or rcParams if set.

I'm resolving this thread - please comment on the latest changes.

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: The version is now pinned and the code got much simpler. Please check the latest changes.

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: Done!

File: pymc3/sampling.py
Comment (OriolAbril) [inline pymc3/sampling.py]: 3 comments on this: 
1. It should be `arviz.rcParams` like in matplotlib
2. I'll open an issue to add this rcParam to ArviZ so it gets added, it currently does not exist
3. Once implemented this if will be taken care of by `from_pymc3` so ti won't be needed, and before being implemented it cannot be triggered. I think this can actually be removed

File: pymc3/sampling.py
Comment (OriolAbril) [inline pymc3/sampling.py]: how about using some `idata_kwargs` or something like this? This way users can change pass dims and coords to `from_pymc3` and even a `prior` dict result of `pm.sample_prior_predictive`

**Review by OriolAbril — CHANGES_REQUESTED** [review]

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: Okay, if I understand this correctly, the setting can't be set? Can you make a 0.8.1 release then?

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: Good idea! I did this in the latest commit.

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: With the `idata_kwargs` setting, this is now user-configurable anyways.

File: RELEASE-NOTES.md
Comment (AlexAndorra) [inline RELEASE-NOTES.md]: Is the beginning of the sentence missing some words or punctuation?

File: pymc3/sampling.py
Comment (AlexAndorra) [inline pymc3/sampling.py]: add "default=False", as above

File: pymc3/sampling.py
Comment (AlexAndorra) [inline pymc3/sampling.py]: "version 3.10", as 4.0 will most probably be for PyMC4

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
Thanks @michaelosthege , looks good 👌 
Just added some comments, and I agree with Oriol's. Almost there 😉 

PS: didn't review the NB yet -- will do soon.
PPS: the tests are failing because of [this issue](https://github.com/arviz-devs/arviz/issues/1202), [fixed on ArviZ master](https://github.com/arviz-devs/arviz/pull/1203) -- Oriol will do a bugfix-release by tomorrow.

**Comment by michaelosthege** [comment]
waiting for #3931 because of pipeline failures

File: RELEASE-NOTES.md
Comment (michaelosthege) [reply RELEASE-NOTES.md]: this line was actually a duplicate. I changed the wording in the deprecations section

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: I removed the specific version number. Let's discuss upcoming version numbers in the chat

**Review by OriolAbril — APPROVED** [review]

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
Thanks @michaelosthege ! 
I'm finishing the review of the example NB -- I have some comments and changes to implement, but not that big. After these changes, I think it will be ready to merge 🍾 
I don't understand how this is related to #3931 though 🤔

File: docs/source/notebooks/api_quickstart.ipynb
Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <ul><li>Can get rid of %matplotlib inline</li><li>Let's replace all seaborn calls by ArviZ ones?</li><li>Maybe organize imports alphabetically, as in other NBs we modified recently?</li></ul><pre class="ql-syntax" spellcheck="false">import arviz as az
import matplotlib.pyplot as plt
import numpy as np
import pymc3 as pm
import theano.tensor as tt
</pre><ul><li>Let's use ArviZ style ;) And in another cell below (this has to do with how MPL sets defaults):</li></ul><pre class="ql-syntax" spellcheck="false">%config InlineBackend.figure_format = 'retina'
az.style.use('arviz-darkgrid')
print('Running on PyMC3 v{}'.format(pm.__version__))
</pre><ul><li>You can look at the NBs in the <a href="https://github.com/pymc-devs/resources" target="_blank">resources repo</a> for an example of how we format NBs.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='1'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>Typo above x2: "<s>sepcified</s> specify a different..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='40'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>Maybe get rid of <code>return_inferencedata</code> here, as it's False, and introduce it below, as you set it to True?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='44'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>Can you add a few lines about what <code>return_inferencedata</code> does, to introduce the difference with the default behavior, and the advantages of using <code>InferenceData</code>? You can link to <a href="https://arviz-devs.github.io/arviz/notebooks/Introduction.html" target="_blank">ArviZ quickstart</a> at the end.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='57'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>No need for "," between "Note" and "that".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='62'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>Let's use ArviZ directly, as you already returned <code>InferenceData</code> ;)</p><p><code>az.plot_trace(idata);</code></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='72'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>Maybe use <code>az.summary</code> instead? More used in practice and contains <code>r_hat</code>  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='74'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p><code>az.plot_forest</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='76'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p><code>az.plot_posterior</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='78'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p><code>az.plot_energy</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='80'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>You can get rid of the Theano warning with this at the top of the NB, with other imports:</p><pre class="ql-syntax" spellcheck="false">import warnings
warnings.simplefilter(action="ignore", category=FutureWarning)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='87'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>You can use <code>az.plot_kde</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='90'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>You can use <code>az.plot_dist</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='93'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>Nice touch using idata ;)</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='97'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>You can use <code>az.plot_ppc</code>  instead</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='100'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>This example would be even nicer if you use the <code>pm.Data</code>  container, to illustrate :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='102'/>

Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <ul><li>After this last cell, can you add a last cell using <a href="https://github.com/rasbt/watermark" target="_blank">the watermark package</a>:</li></ul><p> </p><pre class="ql-syntax" spellcheck="false">%load_ext watermark
%watermark -n -u -v -iv -w
</pre><ul><li>Finally, can you <a href="https://github.com/dfm/black_nbconvert" target="_blank">format the NB with Black</a>: <code>pip install black_nbconvert</code> and then <code>black_nbconvert /path/to/a/notebook.ipynb</code> on the command line -- this will reformat the NB inplace automatically 😉</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='105'/>

**Comment by OriolAbril** [comment]
> I don't understand how this is related to #3931 though thinking

I am now finishing the patch release for ArviZ: 0.8.1. Until then, trace objects without report attribute trigger the bug and crash. In ArviZ 0.8.1 it won't crash (however, warmup handling won't behave correctly which is not too big a deal as it is what has happened always until now :/ )

Again, thanks @michaelosthege

**Comment by OriolAbril** [comment]
ArviZ 0.8.1 released, it should be required instead of 0.8.0

File: docs/source/notebooks/api_quickstart.ipynb
Comment (OriolAbril) [reply docs/source/notebooks/api_quickstart.ipynb]: <p>This will require to add the post_pred dict as <code>posterior_predictive</code>  group in the inference data object. Try something like:</p><p> </p><pre class="ql-syntax" spellcheck="false">idata.concat(az.from_pymc3(posterior_predictive=post_pred), inplace=True)
az.plot_ppc(idata)
# this above will plot the kde for data, to also add the mean as above,
# the ax.axvline has to be kept
</pre><p> </p>

Comment (OriolAbril) [reply docs/source/notebooks/api_quickstart.ipynb]: <p>IIUC, this will only generate posterior predictive samples for the first 500 draws of the first chain. Could <code>samples</code>  be removed to use the default? </p>

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: changed it yesterday to "a future release"

File: docs/source/notebooks/api_quickstart.ipynb
Comment (AlexAndorra) [inline docs/source/notebooks/api_quickstart.ipynb]: <p>As per our discussion, this whole section should probably be replaced by something like:</p><blockquote>It's worth highlighting the design choice we made with&nbsp;<code>logp</code>.&nbsp;As you can see above,&nbsp;<code>logp</code>&nbsp;is being called with arguments, so it's a method of the model instance. More precisely, it puts together a function based on the current state of the model -- or on the state given as argument to&nbsp;<code>logp</code>&nbsp;(see example below).</blockquote><blockquote>For diverse reasons, we assume that a&nbsp;<code>Model</code>&nbsp;instance isn't static. If you need to use&nbsp;<code>logp</code>&nbsp;in an inner loop and it needs to be static, simply use something like&nbsp;<code>logp = model.logp</code>. Here is an example below -- note the caching effect and the speed up:</blockquote>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='10'/>

Comment (michaelosthege) [reply docs/source/notebooks/api_quickstart.ipynb]: <p>I wanted it to be future-proof, given that with this PR, not passing it raises a FutureWarning...</p><p> </p><p>This example actually uses the <code>MultiTrace</code> API to access transformed variable samples.</p>

Comment (michaelosthege) [reply docs/source/notebooks/api_quickstart.ipynb]: <p>I'm removing this comment alltogether, since this is now reported by the logger anyways.</p>

Comment (michaelosthege) [reply docs/source/notebooks/api_quickstart.ipynb]: <p>😂</p>

Comment (michaelosthege) [reply docs/source/notebooks/api_quickstart.ipynb]: <p>yes. typical example for a chain &amp; draw unbiased random thinning convenience method</p>

Comment (michaelosthege) [reply docs/source/notebooks/api_quickstart.ipynb]: <p>At the moment this does not work - <code>pm.sample_posterior_predictive(idata.posterior)</code>  collapses the 3 chains, 1000 draws into 1 chain 3000 draws. <code>arviz.concat</code>  then refuses to concatenate because the dims don't match.</p>

Comment (OriolAbril) [reply docs/source/notebooks/api_quickstart.ipynb]: does using `keep_size=True` in sample_posterior_predictive solve the issue?

Comment (AlexAndorra) [reply docs/source/notebooks/api_quickstart.ipynb]: Ah you're right, let's keep it to `False` explicitly then 👌

Comment (michaelosthege) [reply docs/source/notebooks/api_quickstart.ipynb]: I don'T know what I did wrong, but now it works

**Comment by michaelosthege** [comment]
@AlexAndorra and @OriolAbril thank you for the feedback on the notebook! they were really straightforward to apply.

File: RELEASE-NOTES.md
Comment (AlexAndorra) [inline RELEASE-NOTES.md]: This should say 0.8.2

File: pymc3/sampling.py
Comment (AlexAndorra) [inline pymc3/sampling.py]: Actually, I'm thinking that we should remove the advice about slicing and just nudge users towards using ID, which takes care of warmup and posterior groups automatically, don't you think? 
This will avoid complications in downstream analysis _and_ help users get used to ID, which is probably gonna be the new default anyway. 
cc @OriolAbril

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
Thanks a lot @michaelosthege ! There was a small typo left, and I left a question about how to handle warmups 😉 Going to review the NB now.

File: RELEASE-NOTES.md
Comment (michaelosthege) [reply RELEASE-NOTES.md]: good catch!

File: pymc3/sampling.py
Comment (michaelosthege) [reply pymc3/sampling.py]: I made the change. However, I'd like to remind us that the warmup groups are still in a hidden attribute of the `InferenceData`

**Comment by AlexAndorra** [comment]
@michaelosthege , finished reviewing the NB -- just a few tweaks to add / typos to fix and we'll be good to go 👌

**Comment by michaelosthege** [comment]
The pipelines fails because of https://github.com/arviz-devs/arviz/issues/1210

I could make a workaround for the test case, but I also think having `draws=0` is a valid use case..

**Review by AlexAndorra — COMMENTED** [review]
The NB is a marvel now, thanks a lot for your patience and perseverance on this @michaelosthege !
Waiting for ArviZ 0.8.3 to be released, and I'll merge if tests pass then -- I really, really hope no other edge case will pop up this time 😅 🤞

**Comment by AlexAndorra** [comment]
Problems seem to be fixed on ArviZ master, but before releasing 0.8.3, we need to make sure this really works. To do that, the best would be that Travis tests against ArviZ master, not only against the latest release -- @canyon289 do you know if it's possible and how easy it would be?
Also pinging @twiecki and @ColCarroll, who could have valuable input.

**Comment by OriolAbril** [comment]
Given that @michaelosthege has successfully run the tests locally, I'll merge tomorrow to make release 0.8.3 and hope for the best.

**Comment by AlexAndorra** [comment]
@OriolAbril was emboldened by your successful local tests and released ArviZ 0.8.3 @michaelosthege, so you can push your last changes and I'll merge if tests pass 🤞

**Review by OriolAbril — APPROVED** [review]

**Comment by OriolAbril** [comment]
:tada:

**Review by AlexAndorra — APPROVED** [review]
We've done it, well done guys 👏 Quite a team 💪

**Pull request merged**
